src/ucm/bistro/bistro_ppc64.c

```diff
@@ -0,0 +1,215 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2018.       ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+/* *******************************************************
+ * POWER-PC processors family                            *
+ * ***************************************************** */
+#if defined (__powerpc64__)
+
+#include <sys/mman.h>
+#include <dlfcn.h>
+#include <string.h>
+#include <stdlib.h>
+
+#include <ucm/bistro/bistro.h>
+#include <ucm/bistro/bistro_int.h>
+#include <ucm/util/sys.h>
+#include <ucs/sys/math.h>
+#include <ucs/arch/cpu.h>
+#include <ucs/debug/assert.h>
+
+/* PowerPC instructions used in patching                  */
+/* Reference: "PowerPC User Instruction Set Architecture" */
+
+#define OPCODE(_rt, _rs, _op) \
+     (((_op) << 26) + ((_rt) << 21) + ((_rs) << 16))
+
+#define OP0(_rt, _rs, _ui, _op) \
+     (OPCODE(_rt, _rs, _op) + ((_ui) & 0xffff))
+
+#define MTSPR(_spr, _rs) \
+    (OPCODE(_rs, (_spr) & 0x1f, 31) + (((_spr) & ~UCS_MASK(5)) << 6) + (467 << 1))
+
+#define BCCTR(_bo, _bi, _bh) \
+    (OPCODE(_bo, _bi, 19) + ((_bh) << 11) + (528<<1))
+#define RLDICR(_rt, _rs, _sh, _mb) \
+    (OPCODE(_rs, _rt, 30) + (((_sh) & UCS_MASK(5)) << 11) + ((_sh & ~UCS_MASK(5)) >> 4) + \
+    (((_mb) & UCS_MASK(5)) << 6) + ((_mb) && ~UCS_MASK(5)) + UCS_BIT(2))
+
+#define ADDIS(_rt, _rs, _ui) OP0(_rt, _rs, _ui, 15)
+#define ORI(_rt, _rs, _ui) OP0(_rs, _rt, _ui, 24)
+#define ORIS(_rt, _rs, _ui) OP0(_rs, _rt, _ui, 25)
+
+typedef struct ucm_bistro_base_patch {
+    uint32_t addis;    /* lis     r11,(addr >> 48)     */
+    uint32_t ori1;     /* ori     r11,r11,(addr >> 32) */
+    uint32_t rldicr;   /* rldicr  r11,r11,32,31        */
+    uint32_t oris;     /* oris    r11,r11,(addr >> 16) */
+    uint32_t ori2;     /* ori     r11,r11,addr         */
+} UCS_S_PACKED ucm_bistro_base_patch_t;
+
+typedef struct ucm_bistro_patch {
+    ucm_bistro_base_patch_t super;
+    uint32_t mtspr;    /* mtctr   r11 */
+    uint32_t bcctr;    /* bcctr       */
+} UCS_S_PACKED ucm_bistro_patch_t;
+
+struct ucm_bistro_restore_point {
+    void                    *entry;
+    void                    *hook;
+    ucm_bistro_base_patch_t hook_patch;
+    void                    *func;
+    ucm_bistro_patch_t      func_patch;
+};
+
+static void ucm_bistro_fill_base_patch(ucm_bistro_base_patch_t *patch,
+                                       uint32_t reg, uintptr_t value)
+{
+    ucs_assert(patch != NULL);
+
+    UCS_STATIC_ASSERT(ucs_offsetof(ucm_bistro_base_patch_t, addis)  == 0);
+    UCS_STATIC_ASSERT(ucs_offsetof(ucm_bistro_base_patch_t, ori1)   == 4);
```

Comment:
do we really need all those static asserts? every instruction is same length anyway...

Reply:
right, but if somebody decide to update patch structure (exchange operations, for example) - he will think about this twice :)

Reply:
static assert is usually for stuff which are not directly deduced from looking at the code, for example total size of a struct
for thing which can be easily observed in the code, such as size of primitive field of field order this is redundant IMHO

Reply:
ok, removed